File: app/src/main/java/jp/ac/u_tokyo/t/seo/station/app/SettingFragment.java
package jp.ac.u_tokyo.t.seo.station.app;

import android.content.Context;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.NumberPicker;
import android.widget.SeekBar;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.ToggleButton;

import java.util.Locale;

import jp.ac.u_tokyo.t.seo.customdialog.CompatFragment;
import jp.ac.u_tokyo.t.seo.customview.CustomNumberPicker;

/**
 * @author Seo-4d696b75
 * @version 2018/04/25.
 */

public class SettingFragment extends CompatFragment implements DataDialog.DataUpdateResult{

    private MainActivity mContext;
    private StationService mService;

    private TextView mTextDataVersion;

    @Override
    public void onAttach(Context context){
        super.onAttach(context);
        if ( context instanceof MainActivity ){
            mContext = (MainActivity)context;
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle state){
        super.onCreateView(inflater, container, state);
        return inflater.inflate(R.layout.fragment_setting, container, false);
    }

    @Override
    public void onViewCreated(final View view, Bundle state){
        mContext.getService(new MainActivity.OnServiceConnectedListener(){
            @Override
            public void onServiceConnected(StationService service){
                mService = service;
                init(view);
            }
        });

    }

    private void init(View view){
        final Switch notification = (Switch)view.findViewById(R.id.switchNotify);
        notification.setChecked(mService.isNotifyUpdate());

        final Switch forceNotify = (Switch)view.findViewById(R.id.switchForceNotify);
        forceNotify.setChecked(mService.isNotifyForce());

        final Switch stationaryNotify = (Switch)view.findViewById(R.id.switchNotifyStationary);
        stationaryNotify.setChecked(mService.isKeepNotification());

        notification.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener(){
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked){
                mService.setNotify(isChecked);
                if ( !isChecked ){
                    forceNotify.setChecked(false);
                    stationaryNotify.setChecked(false);
                }
            }
        });
        forceNotify.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener(){
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked){
                mService.setForceNotify(isChecked);
                if ( isChecked && !mService.isNotifyUpdate() ){
                    notification.setChecked(true);
                }
            }
        });
        stationaryNotify.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener(){
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked){
                mService.setKeepNotification(isChecked);
                if ( isChecked && !mService.isNotifyUpdate() ){
                    notification.setChecked(true);
                }
            }
        });
        Switch prefecture = (Switch)view.findViewById(R.id.switchDisplayPrefecture);
        prefecture.setChecked(mService.isNotifyPrefecture());
        prefecture.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener(){
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked){
                mService.setNotifyPrefecture(isChecked);
            }
        });

        CustomNumberPicker interval = (CustomNumberPicker)view.findViewById(R.id.numberInterval);
        interval.setDisplayedValue(mService.getInterval());
        interval.setOnValueChangedListener(new NumberPicker.OnValueChangeListener(){
            @Override
            public void onValueChange(NumberPicker picker, int oldVal, int newVal){
                mService.setInterval(newVal, true);
            }
        });
        CustomNumberPicker radar = (CustomNumberPicker)view.findViewById(R.id.numberRadar);
        radar.setDisplayedValue(mService.getRadarNum());
        radar.setOnValueChangedListener(new NumberPicker.OnValueChangeListener(){
            @Override
            public void onValueChange(NumberPicker picker, int oldVal, int newVal){
                mService.setRadarNum(newVal);
            }
        });
        final CustomNumberPicker meter = (CustomNumberPicker)view.findViewById(R.id.numberVibrateMeter);
        meter.setDisplayedValue(mService.getVibrateMeter());
        meter.setEnabled(mService.isVibrate() && mService.isVibrateApproach());
        meter.setOnValueChangedListener(new NumberPicker.OnValueChangeListener(){
            @Override
            public void onValueChange(NumberPicker picker, int oldVal, int newVal){
                mService.setVibrateMeter(newVal);
            }
        });
        final Switch vibrateApproach = (Switch)view.findViewById(R.id.switchApproachVibrate);
        vibrateApproach.setChecked(mService.isVibrateApproach());
        vibrateApproach.setEnabled(mService.isVibrate());
        vibrateApproach.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener(){
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked){
                mService.setVibrateApproach(isChecked);
                meter.setEnabled(isChecked);
            }
        });
        Switch vibration = (Switch)view.findViewById(R.id.switchVibrate);
        vibration.setChecked(mService.isVibrate());
        vibration.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener(){
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked){
                mService.setVibrate(isChecked);
                vibrateApproach.setEnabled(isChecked);
                meter.setEnabled(isChecked && mService.isVibrateApproach());
            }
        });
        mTextDataVersion = (TextView)view.findViewById(R.id.textDataVersion);
        showDataVersion();
        final View dataUpdate = view.findViewById(R.id.imageDataVersion);
        dataUpdate.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View v){
                mService.checkDataVersion(new StationService.DataVersionResult(){
                    @Override
                    public void onUpToDate(){
                        Toast.makeText(mContext, getString(R.string.data_version_latest), Toast.LENGTH_LONG).show();
                    }

                    @Override
                    public void onLatestFound(long version, String path, String size){
                        DataDialog dialog = DataDialog.getInstance(version, path, size, false);
                        dialog.show(getCompatFragmentManager(), null);
                    }

                    @Override
                    public void onCheckFailure(){
                        Toast.makeText(mContext, getString(R.string.data_version_check_fail), Toast.LENGTH_LONG).show();
                    }
                });
            }
        });
        dataUpdate.setOnLongClickListener(new View.OnLongClickListener(){
            @Override
            public boolean onLongClick(View v){
                DataDialog dialog = DataDialog.getInstance();
                dialog.show(getCompatFragmentManager(), null);
                return true;
            }
        });

        Switch night = (Switch)view.findViewById(R.id.switchNight);
        night.setChecked(mService.isNightMood());
        night.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener(){
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked){
                mService.setNightMood(isChecked);
            }
        });
        SeekBar brightness = (SeekBar)view.findViewById(R.id.seekBrightness);
        final View screen = view.findViewById(R.id.viewSampleTextScreen);
        brightness.setMax(255);
        brightness.setProgress(mService.getBrightness());
        brightness.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener(){
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser){
                Drawable back = new ColorDrawable((255 - progress) << 24);
                screen.setBackground(back);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar){

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar){
                mService.setBrightness(seekBar.getProgress());
            }
        });
    }


    @Override
    public void onUpdateRequired(boolean required, long version, String path){
        if ( required ){
            DataDialog dialog = DataDialog.getInstance(version, path);
            dialog.show(getCompatFragmentManager(), null);
        }
    }

    @Override
    public void onDateUpdate(long version, boolean success){
        showDataVersion();
        if ( success ){
            Toast.makeText(getContext(), "データの更新が完了しました", Toast.LENGTH_LONG).show();
        }else{
            Toast.makeText(getContext(), "データの更新に失敗しました", Toast.LENGTH_LONG).show();
        }
    }

    private void showDataVersion(){
        if ( mService != null ){
            String message = String.format(Locale.US, getString(R.string.data_version_format), mService.getDataVersion());
            mTextDataVersion.setText(message);
        }
    }

    @Override
    public void onStop(){
        super.onStop();
        if ( mService != null ){
            mService.saveSetting();
        }
    }

    @Override
    public void onDestroy(){
        super.onDestroy();
        mService = null;
        mContext = null;
        mTextDataVersion = null;
    }

}

File: app/src/main/java/jp/seo/station/ekisagasu/ui/SettingFragment.kt
package jp.seo.station.ekisagasu.ui

import android.content.Context
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.widget.SwitchCompat
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import dagger.hilt.android.AndroidEntryPoint
import jp.seo.android.widget.CustomNumberPicker
import jp.seo.station.ekisagasu.R
import jp.seo.station.ekisagasu.repository.AppStateRepository
import jp.seo.station.ekisagasu.utils.TIME_PATTERN_DATETIME
import jp.seo.station.ekisagasu.utils.formatTime
import jp.seo.station.ekisagasu.viewmodel.ActivityViewModel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * @author Seo-4d696b75
 * @version 2021/01/20.
 */
@AndroidEntryPoint
class SettingFragment : AppFragment() {

    private val activityViewModel: ActivityViewModel by lazy {
        ActivityViewModel.getInstance(
            requireActivity(),
            requireContext(),
            stationRepository,
            userRepository
        )
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_setting, container, false)
    }

    @Inject
    lateinit var appStateRepository: AppStateRepository

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val notify = view.findViewById<SwitchCompat>(R.id.switch_notification)
        userRepository.isNotify.value?.let { notify.isChecked = it }
        val forceNotify = view.findViewById<SwitchCompat>(R.id.switch_force_notify)
        userRepository.isNotifyForce.value?.let { forceNotify.isChecked = it }
        val keepNotification = view.findViewById<SwitchCompat>(R.id.switch_keep_notification)
        userRepository.isKeepNotification.value?.let { keepNotification.isChecked = it }

        notify.setOnCheckedChangeListener { buttonView, isChecked ->
            userRepository.isNotify.value = isChecked
            if (!isChecked) {
                forceNotify.isChecked = false
                keepNotification.isChecked = false
            }
        }
        forceNotify.setOnCheckedChangeListener { buttonView, isChecked ->
            userRepository.isNotifyForce.value = isChecked
            if (isChecked) {
                notify.isChecked = true
            }
        }
        keepNotification.setOnCheckedChangeListener { buttonView, isChecked ->
            userRepository.isKeepNotification.value = isChecked
            if (isChecked) {
                notify.isChecked = true
            }
        }

        val prefecture = view.findViewById<SwitchCompat>(R.id.switch_display_prefecture)
        userRepository.isNotifyPrefecture.value?.let { prefecture.isChecked = it }
        prefecture.setOnCheckedChangeListener { buttonView, isChecked ->
            userRepository.isNotifyPrefecture.value = isChecked
        }

        val interval = view.findViewById<CustomNumberPicker>(R.id.number_time_interval)
        userRepository.gpsUpdateInterval.value?.let { interval.displayedValue = it }
        interval.setOnValueChangedListener { picker, oldVal, newVal ->
            userRepository.gpsUpdateInterval.value = newVal
        }

        val radar = view.findViewById<CustomNumberPicker>(R.id.number_radar)
        userRepository.searchK.value?.let { radar.displayedValue = it }
        radar.setOnValueChangedListener { picker, oldVal, newVal ->
            userRepository.searchK.value = newVal
        }

        val vibrate = view.findViewById<SwitchCompat>(R.id.switch_vibrate)
        val vibrateApproach = view.findViewById<SwitchCompat>(R.id.switch_vibrate_approach)
        val vibrateDistance = view.findViewById<CustomNumberPicker>(R.id.number_vibrate_meter)
        userRepository.isVibrateApproach.value?.let { vibrateApproach.isChecked = it }
        userRepository.vibrateDistance.value?.let { vibrateDistance.displayedValue = it }
        userRepository.isVibrate.value?.let {
            vibrate.isChecked = it
            vibrateApproach.isEnabled = it
            vibrateDistance.isEnabled = it
        }
        vibrate.setOnCheckedChangeListener { buttonView, isChecked ->
            userRepository.isVibrate.value = isChecked
            vibrateApproach.isEnabled = isChecked
            vibrateDistance.isEnabled = isChecked && vibrateApproach.isChecked
        }
        vibrateApproach.setOnCheckedChangeListener { buttonView, isChecked ->
            userRepository.isVibrateApproach.value = isChecked
            vibrateDistance.isEnabled = isChecked
        }
        vibrateDistance.setOnValueChangedListener { picker, oldVal, newVal ->
            userRepository.vibrateDistance.value = newVal
        }

        val night = view.findViewById<SwitchCompat>(R.id.switch_night)

        appStateRepository.nightMode
            .flowWithLifecycle(lifecycle)
            .onEach { night.isChecked = it }
            .launchIn(lifecycleScope)

        night.setOnCheckedChangeListener { _, isChecked ->
            lifecycleScope.launch {
                appStateRepository.setNightMode(isChecked)
            }
        }

        val nightTimeout = view.findViewById<Spinner>(R.id.spinner_night_mode_switch)
        context?.let { ctx ->
            val values = ctx.resources.getIntArray(R.array.night_mode_switch_timeout).toTypedArray()
            nightTimeout.adapter = NightModeTimeoutAdapter(ctx, values)
            userRepository.nightModeTimeout.value?.let { timeout ->
                val index = values.indexOfFirst { it == timeout }
                if (index >= 0) nightTimeout.setSelection(index)
            }
            nightTimeout.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(
                    parent: AdapterView<*>?,
                    view: View?,
                    position: Int,
                    id: Long
                ) {
                    parent?.getItemAtPosition(position)?.let {
                        if (it is Int) {
                            userRepository.nightModeTimeout.value = it
                        }
                    }
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        }

        val brightness = view.findViewById<SeekBar>(R.id.seek_brightness)
        val screen = view.findViewById<View>(R.id.view_sample_brightness)
        brightness.max = 255
        brightness.min = OverlayViewHolder.MIN_BRIGHTNESS
        userRepository.brightness.value?.let {
            brightness.progress = it
            screen.background = ColorDrawable((255 - it).shl(24))
        }
        brightness.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                val back = ColorDrawable((255 - progress).shl(24))
                screen.background = back
            }

            override fun onStartTrackingTouch(seekBar: SeekBar?) {
            }

            override fun onStopTrackingTouch(seekBar: SeekBar?) {
                userRepository.brightness.value = brightness.progress
            }

        })

        val version = view.findViewById<TextView>(R.id.text_data_version)
        val timestamp = view.findViewById<TextView>(R.id.text_data_timestamp)
        val checkVersion = view.findViewById<Button>(R.id.button_update_data)
        stationRepository.dataVersion.observe(viewLifecycleOwner) {
            it?.let { data ->
                version.text = String.format(
                    "%s: %d",
                    getString(R.string.setting_mes_data_version),
                    data.version
                )
                timestamp.text = String.format(
                    "%s: %s",
                    getString(R.string.setting_mes_data_timestamp),
                    formatTime(TIME_PATTERN_DATETIME, data.timestamp)
                )
            }
        }
        checkVersion.setOnClickListener {
            activityViewModel.checkLatestData(requireContext())
        }

    }
}

class NightModeTimeoutAdapter(
    context: Context,
    values: Array<Int>
) : ArrayAdapter<Int>(context, 0, values) {

    private val inflater = LayoutInflater.from(context)

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view = convertView ?: inflater.inflate(android.R.layout.simple_spinner_item, null)
        val value = getItem(position)
        if (value != null && view is TextView) {
            view.text = if (value == 0) {
                context.getString(R.string.setting_mes_night_switch_always)
            } else {
                String.format(
                    context.getString(R.string.setting_mes_night_switch),
                    formatTime(context, value)
                )
            }
        }
        return view
    }

    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view =
            convertView ?: inflater.inflate(android.R.layout.simple_spinner_dropdown_item, null)
        val value = getItem(position)
        if (value != null && view is TextView) {
            view.text = if (value == 0) {
                context.getString(R.string.setting_mes_night_switch_always)
            } else {
                formatTime(context, value)
            }
        }
        return view
    }
}
